ocaml/xapi/xapi_sr.ml

```diff
@@ -778,13 +783,40 @@ let scan ~__context ~sr =
               Db.VDI.get_records_where ~__context
                 ~expr:(Eq (Field "SR", Literal sr'))
             in
+            let string_of_l lst =
+              List.map (fun (_, v) -> v.vDI_uuid) lst |> String.concat ","
+            in
             (* It is sufficient to just compare the refs in two db_vdis, as this
                is what update_vdis uses to determine what to delete *)
             let vdis_ref_equal db_vdi1 db_vdi2 =
-              Listext.List.set_difference (List.map fst db_vdi1)
-                (List.map fst db_vdi2)
-              = []
+              let refs1 = RefSet.of_list (List.map fst db_vdi1) in
+              let refs2 = RefSet.of_list (List.map fst db_vdi2) in
+              (* VDIs that are in db_vdi1 but not in db_vdi2 *)
+              let vdis_removed = RefSet.diff refs1 refs2 in
+              (* VDIs that are in not in db_vdi1 but db_vdi2 *)
+              let vdis_added = RefSet.diff refs2 refs1 in
+
+              ( if not (RefSet.is_empty vdis_removed) then
+                  let removed =
```

Comment:
If we report the Refs, we could simply turn the set into strings and no filtering would be required. Same below.